daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/PartialTransaction.scala

```diff
@@ -180,16 +180,33 @@ private[lf] object PartialTransaction {
   *                 change to a transaction.
-  *  @param keys A local store of the contract keys. Note that this contains
-  *              info both about relative and contract ids. We must
-  *              do this because contract ids can be archived as
-  *              part of execution, and we must record these archivals locally.
-  *              Note: it is important for keys that we know to not be present
-  *              to be present as [[None]]. The reason for this is that we must
-  *              record the "no key" information for contract ids that
-  *              we archive. This is not an optimization and is required for
-  *              correct semantics, since otherwise lookups for keys for
-  *              locally archived contract ids will succeed wrongly.
+  *  @param keys A local store of the contract keys used for lookups and fetches by keys
+  *              (including exercise by key). Each of those operations will be resolved
+  *              against against this map first. Only if there is no entry in here
+  *              (but not if there is an entry mapped to None), will we ask the ledger.
+  *
+  *              This map is mutated by the following operations:
+  *              1. fetch-by-key/lookup-by-key/exercise-by-key will insert an
+  *                 an entry in the map if there wasn’t already one (i.e., if they queried the ledger).
+  *              2. ACS mutating operations if the corresponding contract has a key. Specifically,
+  *                 2.1. A create will set the corresponding map entry to Some(cid) if the contract has a key.
+  *                 2.2. A consuming choice will set the corresponding map entry to None if the contract has a key.
+  *
+  *              On a rollback, we restore the state at the beginning of the rollback. Note that
+  *              This means that at the moment, we will query (by-key) for a global contract again even if
+  *              we queried it in a rollback node already. We can cache this to improve
+  *              performance in the future.
+  *
+  *              In non-uck (Unique Contract Key) mode, i.e., Canton this is all we need to do.
+  *              In uck mode (not yet implemented), Speedy will detect and reject duplicate keys.
+  *              Only the operations that mutate `keys` will be taken into account for that.
+  *              Specifically, non-ACS operations that are not by-key (fetch, non-consuming exercises)
+  *              can never result in a failure due to duplicate keys.
+  *              For duplicate key detection, rollbacks roll back all writes but not reads.
+  *              So a global fetch by key in a rollback node will still conflict with
+  *              a create outside of the rollback node. And a create, comes with an implicit
+  *              NoSuchKey read so a create in a rollback node will conflict with a global fetch
+  *              by key outside of the rollback.
```

Comment:
Here's my suggestions of how to model keys. We have three kinds of operations on the `keys` map:
1. lookup of an entry in the `keys` map
2. caching a ledger response upon a `ResultNeedKey`
3. inserting/updating a key

A rollback node reverts all inserts/updates by its transitive children, but keeps their cache updates. This ensures that Daml engine queries the ledger for every key at most once during one interpretation run.

The operations are used as follows in non-uck mode:

1. A `Create` node of a contract with a key `k` upserts `keys` at `k` to the new contract ID. It does not care what was stored previously at `k`.
2. `Fetch` and non-consuming `Exercise` do not lookup or cache anything in `keys`.
3. `FetchByKey` and `ExerciseByKey` lookup the key `k` in keys `keys`. If they find an entry for `k`, that's the contract they use. If the don't find an entry, they ask the ledger with `ResultNeedKey` and cache the ledger's response in `keys`. If the `ExerciseByKey` is consuming, it then updates the key `k` to `None`.
4. A consuming `Exercise` on a contract `cid` with a key `k` checks whether `keys` associates `k` with `Some cid`. If so, it updates the association to `None`. Otherwise, `keys` remains as is.

Reply:
Seems like a sensible model. To spell out the diff to the current state explicitly:
1. Cache across rollback nodes making sure the engine only ever looks up a key once. I’d argue the ledger is still broken if it returns two different results during execution but it seems like a very sensible expectation on the engine nevertheless.
2. Change archives to only set a key to None if it would have been returned by a directly preceding `fetchByKey`.

Reply:
Thanks for the concrete model @andreaslochbihler-da. I agree with the diff that you @cocreature list.

@both: I'm no longer convinced, that resolving a key at most once is the right action in non-uck mode. The reason is that it implies that composing two working commands into a single one can lead to command interpretation failures.
```
-- setup
cid1 <- submit p $ create (Keyed p)
cid2 <- submit p $ create (Keyed p)
-- this works
submit p $ archiveByKey p
submit p $ archiveByKey p
-- using the composed version instead of the above two commands would fail
submit p $ do
  archiveByKey p
  archiveByKey p
```

Questions:
1. I believe our current system satisfies this composition property for uck mode, do you agree?
2. Do you believe we should keep this property?
3. If yes, how do we want to define non-uck key semantics so that we can efficiently implement it?

Reply:
1. agreed
2. seems like a very sensible and desirable property, phrasing it differently: The difference between batching commands into a single submission is transactionality and nothing else.
3. adding a set of exclusions to ResultNeedsKey seems easy on the engine side and hopefully reasonably efficient.

That said, overall I’m increasingly hesitant to tie multi-key semantics to doing duplicate key checks in the engine. The latter is a problem we have right now that results in actual bugs. I want to get rid of the 4 different implementations in the Daml repo sooner rather than later and I don’t think we should rush multi-key semantics. So I think we should have two related but orthogonal discussions:

1. What does an uck-mode for the engine reject? I think we agree on by-key operations and create. The open one seems to be `archive`. I think I still prefer @andreaslochbihler-da’s proposal here.
2. What should multi-key semantics be?

Reply:
1. With a UCK ledger (e.g., Canton in UCK mode), the second `submit p $ archiveByKey p` will fail. (IMO there's little point to run Daml engine in UCK mode on a non-uck ledger.) So why do you think that this works?
2. We violate the compositionality property already in other places. For example, in a multi-domain setup, transaction 1 could be routed to domain D1 and transaction 2 to domain D2, but there's no domain on which the combination of both transactions could be executed. (This problem will remain even with cross-domain transactions.)

The composition property is actually preserved with the current design under @oggy- 's fetch-Some semantics, which is what we guarantee to Daml developers. Under this semantics, the second `archiveByKey p` may fail even when we submit the transactions individually. And we do not promise to our users that aggregating several transactions into one has the same behaviour w.r.t. resolving the non-determinism around fetchSome.

Reply:
I think the example @meiersi-da provided was intended to be run in non-uck. With the semantics you proposed, the first one will succeed. The second one will not.

You are correct, that in terms of guarantees the first one could also fail but I think there is still a useful difference between "this could fail given the guarantees we provide" and "this will fail on an honest submitter".

Reply:
@cocreature : I agree to the separation of solving uck semantics of the engine and non-uck semantics.

As for uck-semantics, I'm in agreement with @andreaslochbihler-da 's proposal.

Reply:
Resolving since we agreed to postpone non-uck. Will implement the proposal for uck in separate PRs.